File: 56-merge-intervals/56-merge-intervals.js
/**
 * @param {number[][]} intervals
 * @return {number[][]}
 */
var merge = function(intervals) {
    intervals.sort((a,b) => {
        if(a[0] < b[0]) {
            return -1
        }
        else if(a[0] == b[0] && a[1] < b[1]) {
            return -1
        } else {
            return 1
        }
    })
    //console.log(intervals)
    let mergedIntervals = []
    if(intervals.length <= 0) return mergedIntervals
    let tempinterval = intervals[0]
    for(let interval of intervals) {
        if(interval[0] <= tempinterval[1]) {
            tempinterval[0] = Math.min(interval[0], tempinterval[0])
            tempinterval[1] = Math.max(interval[1], tempinterval[1])
        } else {
            mergedIntervals.push(tempinterval)
            tempinterval = interval
        }
    }
    mergedIntervals.push(tempinterval)
    return mergedIntervals
};

File: 56-merge-intervals/56-merge-intervals.java
class Solution {
    public int[][] merge(int[][] intervals) {
        int n = intervals.length;
        Arrays.sort(intervals, (a,b) -> {
            if(a[0] == b[0]) return a[1] - b[1];
            return a[0]- b[0];
        });
        
        ArrayList<int[]> al = new ArrayList<int[]>();
        al.add(intervals[0]);
        for(int i = 1; i < n; i++) {
            int[] last = al.get(al.size() - 1);
            int[] curr = intervals[i];
            if(curr[0] > last[1]) {
                al.add(curr);
            } else {
                al.remove(al.size() - 1);
                int min = Math.min(last[0], curr[0]);
                int max = Math.max(last[1], curr[1]);
                al.add(new int[]{min, max});
            }
        }
        //System.out.println(al.size());
        int[][] ans = new int[al.size()][2];
        int i = 0;
        while(i < al.size()) {
            ans[i] = al.get(i);
            i++;
        }
        
        return ans;
        
    }
}